src/lib.rs

```diff
@@ -17,244 +16,178 @@
-    fn absorb(&mut self, input: &impl Absorbable<F>);
+    fn absorb(&mut self, input: &impl Absorb);
 
     /// Squeeze `num_bytes` bytes from the sponge.
     fn squeeze_bytes(&mut self, num_bytes: usize) -> Vec<u8>;
 
+    /// Squeeze `num_bits` bits from the sponge.
+    fn squeeze_bits(&mut self, num_bits: usize) -> Vec<bool>;
+
     /// Squeeze `sizes.len()` field elements from the sponge, where the `i`-th element of
     /// the output has size `sizes[i]`.
-    fn squeeze_field_elements_with_sizes(&mut self, sizes: &[FieldElementSize]) -> Vec<F>;
-
-    /// Squeeze `num_elements` field elements from the sponge.
-    fn squeeze_field_elements(&mut self, num_elements: usize) -> Vec<F> {
-        self.squeeze_field_elements_with_sizes(
-            vec![FieldElementSize::Full; num_elements].as_slice(),
-        )
-    }
-}
-
-/// An interface for objects that can be absorbed by a `CryptographicSponge`.
-pub trait Absorbable<F: PrimeField> {
-    /// Converts the object into a list of bytes that can be absorbed by a `CryptographicSponge`.
-    fn to_sponge_bytes(&self) -> Vec<u8>;
+    ///
+    /// If the implementation is field-based, to squeeze native field elements,
+    /// call `self.squeeze_native_field_elements` instead.
+    ///
+    /// TODO: Support general Field.
+    ///
+    /// Note that when `FieldElementSize` is `FULL`, the output is not strictly uniform. Output
+    /// space is uniform in \[0, 2^{F::MODULUS_BITS - 1}\]
+    fn squeeze_field_elements_with_sizes<F: PrimeField>(
+        &mut self,
+        sizes: &[FieldElementSize],
+    ) -> Vec<F> {
+        if sizes.len() == 0 {
+            return Vec::new();
+        }
 
-    /// Converts the object into field elements that can be absorbed by a `CryptographicSponge`.
-    fn to_sponge_field_elements(&self) -> Vec<F>;
-}
+        let mut total_bits = 0usize;
+        for size in sizes {
+            total_bits += size.num_bits::<F>();
+        }
 
-impl<F: PrimeField> Absorbable<F> for u8 {
-    fn to_sponge_bytes(&self) -> Vec<u8> {
-        vec![*self]
-    }
+        let bits = self.squeeze_bits(total_bits);
+        let mut bits_window = bits.as_slice();
+
+        let mut output = Vec::with_capacity(sizes.len());
+        for size in sizes {
+            let num_bits = size.num_bits::<F>();
+            let nonnative_bits_le: Vec<bool> = bits_window[..num_bits].to_vec();
+            bits_window = &bits_window[num_bits..];
+
+            let nonnative_bytes = nonnative_bits_le
+                .chunks(8)
+                .map(|bits| {
+                    let mut byte = 0u8;
+                    for (i, &bit) in bits.into_iter().enumerate() {
+                        if bit {
+                            byte += 1 << i;
+                        }
+                    }
+                    byte
+                })
+                .collect::<Vec<_>>();
+
+            output.push(F::from_random_bytes(nonnative_bytes.as_slice()).unwrap());
```

Comment:
We should use `from_bytes_mod_order`, not `from_random_bytes`.

Reply:
fixed